File: backend/services/telegramService.js
// backend/services/telegramService.js - Service Telegram MAKERHUB V1
'use strict';

const firebaseService = require('./firebaseService');

/**
 * Service Telegram pour la gestion des canaux et membres
 * Note: L'intégration bot complète est dans le service Python
 */
class TelegramService {
  constructor() {
    this.botToken = process.env.TELEGRAM_TOKEN;
    this.botUsername = process.env.BOT_USERNAME || '@Makerhubsub_bot';
    this.pythonServiceUrl = process.env.PYTHON_SERVICE_URL || 'http://localhost:5001';
  }

  // ==================== CHANNEL MANAGEMENT ====================

  /**
   * Connecter un canal Telegram à une page
   */
  async connectChannel(pageId, channelLink) {
    try {
      // Valider le lien
      const channelInfo = this.parseChannelLink(channelLink);
      if (!channelInfo) {
        throw new Error('Invalid Telegram channel link');
      }

      // Mettre à jour la page
      await firebaseService.updateLandingPage(pageId, {
        telegramChannelId: channelInfo.id || channelInfo.username,
        telegramChannelLink: channelLink,
        telegramConnected: true
      });

      return {
        success: true,
        channelId: channelInfo.id || channelInfo.username,
        channelLink: channelLink
      };

    } catch (error) {
      console.error('❌ Connect channel error:', error);
      throw error;
    }
  }

  /**
   * Déconnecter un canal Telegram
   */
  async disconnectChannel(pageId) {
    try {
      await firebaseService.updateLandingPage(pageId, {
        telegramChannelId: null,
        telegramChannelLink: null,
        telegramConnected: false
      });

      return { success: true };

    } catch (error) {
      console.error('❌ Disconnect channel error:', error);
      throw error;
    }
  }

  /**
   * Parser un lien de canal Telegram
   */
  parseChannelLink(link) {
    if (!link) return null;

    // Format: https://t.me/+hash (invite privé)
    const privateMatch = link.match(/t\.me\/\+([a-zA-Z0-9_-]+)/);
    if (privateMatch) {
      return { type: 'private', id: privateMatch[1] };
    }

    // Format: https://t.me/username (public)
    const publicMatch = link.match(/t\.me\/([a-zA-Z0-9_]+)/);
    if (publicMatch) {
      return { type: 'public', username: publicMatch[1] };
    }

    // Format: @username
    const usernameMatch = link.match(/^@([a-zA-Z0-9_]+)$/);
    if (usernameMatch) {
      return { type: 'public', username: usernameMatch[1] };
    }

    return null;
  }

  // ==================== MEMBER MANAGEMENT ====================

  /**
   * Ajouter un membre à un canal (via service Python)
   */
  async addMemberToChannel(pageId, telegramUserId, email) {
    try {
      const response = await fetch(`${this.pythonServiceUrl}/api/telegram/add-member`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          page_id: pageId,
          telegram_user_id: telegramUserId,
          email: email
        })
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.message || 'Failed to add member');
      }

      const result = await response.json();

      // Enregistrer dans Firebase
      await firebaseService.addTelegramMember({
        pageId,
        telegramUserId,
        email,
        inviteLink: result.invite_link
      });

      return result;

    } catch (error) {
      console.error('❌ Add member to channel error:', error);
      throw error;
    }
  }

  /**
   * Retirer un membre d'un canal (via service Python)
   */
  async removeMemberFromChannel(pageId, telegramUserId) {
    try {
      const response = await fetch(`${this.pythonServiceUrl}/api/telegram/remove-member`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          page_id: pageId,
          telegram_user_id: telegramUserId
        })
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.message || 'Failed to remove member');
      }

      return await response.json();

    } catch (error) {
      console.error('❌ Remove member from channel error:', error);
      throw error;
    }
  }

  /**
   * Récupérer les membres d'une page
   */
  async getMembers(pageId) {
    return firebaseService.getTelegramMembersByPage(pageId);
  }

  /**
   * Compter les membres d'une page
   */
  async getMemberCount(pageId) {
    const members = await this.getMembers(pageId);
    return members.length;
  }

  // ==================== INVITE LINKS ====================

  /**
   * Créer un lien d'invitation (via service Python)
   */
  async createInviteLink(pageId, options = {}) {
    try {
      const response = await fetch(`${this.pythonServiceUrl}/api/telegram/create-invite`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          page_id: pageId,
          expire_hours: options.expireHours || 24,
          member_limit: options.memberLimit || 1
        })
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.message || 'Failed to create invite link');
      }

      return await response.json();

    } catch (error) {
      console.error('❌ Create invite link error:', error);
      throw error;
    }
  }

  // ==================== HEALTH CHECK ====================

  /**
   * Vérifier la connexion avec le service Python
   */
  async checkPythonService() {
    try {
      const response = await fetch(`${this.pythonServiceUrl}/api/health`);
      if (response.ok) {
        return { status: 'ok', service: 'python' };
      }
      return { status: 'error', service: 'python' };
    } catch (error) {
      return { status: 'unreachable', service: 'python', error: error.message };
    }
  }

  /**
   * Vérifier la configuration du bot
   */
  async checkBotConfig() {
    return {
      configured: !!this.botToken,
      botUsername: this.botUsername,
      pythonServiceUrl: this.pythonServiceUrl
    };
  }
}

// Export singleton
module.exports = new TelegramService();

File: telegram/services/TelegramService.py
# telegram/services/TelegramService.py
"""
Service Telegram pour MAKERHUB V1
Gestion des canaux, membres et invitations
"""

import os
import logging
from datetime import datetime, timedelta

logger = logging.getLogger(__name__)


class TelegramService:
    """Service pour gérer les interactions Telegram"""
    
    def __init__(self):
        self.bot_token = os.getenv('TELEGRAM_TOKEN')
        self.bot_username = os.getenv('BOT_USERNAME', '@Makerhubsub_bot')
        self.api_id = os.getenv('TELEGRAM_API_ID')
        self.api_hash = os.getenv('TELEGRAM_API_HASH')
        self._bot = None
    
    def _get_bot(self):
        """Lazy loading du bot Telegram"""
        if not self._bot and self.bot_token:
            try:
                from telegram import Bot
                self._bot = Bot(token=self.bot_token)
            except Exception as e:
                logger.error(f"Failed to initialize bot: {e}")
        return self._bot
    
    # ==================== HEALTH CHECK ====================
    
    def check_bot_sync(self):
        """Vérifie la connexion au bot (synchrone)"""
        try:
            if not self.bot_token:
                return {'status': 'error', 'message': 'No bot token configured'}
            
            return {
                'status': 'ok',
                'service': 'telegram',
                'bot_username': self.bot_username,
                'configured': True
            }
        except Exception as e:
            logger.error(f"Bot check error: {e}")
            return {'status': 'error', 'message': str(e)}
    
    # ==================== INVITE LINKS ====================
    
    def create_invite_link_sync(self, channel_id, expire_hours=24, member_limit=1):
        """Crée un lien d'invitation (synchrone)"""
        try:
            import asyncio
            return asyncio.run(self._create_invite_link_async(channel_id, expire_hours, member_limit))
        except Exception as e:
            logger.error(f"Create invite link error: {e}")
            return {'error': str(e)}
    
    async def _create_invite_link_async(self, channel_id, expire_hours=24, member_limit=1):
        """Crée un lien d'invitation (async)"""
        try:
            from telegram import Bot
            
            bot = Bot(token=self.bot_token)
            expire_date = datetime.now() + timedelta(hours=expire_hours)
            
            invite_link = await bot.create_chat_invite_link(
                chat_id=channel_id,
                expire_date=expire_date,
                member_limit=member_limit
            )
            
            logger.info(f"✅ Invite link created for channel {channel_id}")
            
            return {
                'success': True,
                'invite_link': invite_link.invite_link,
                'expire_date': expire_date.isoformat()
            }
            
        except Exception as e:
            logger.error(f"Create invite link async error: {e}")
            return {'error': str(e)}
    
    # ==================== MEMBER MANAGEMENT ====================
    
    def add_member_to_channel_sync(self, page_id, telegram_user_id, email=''):
        """Ajoute un membre au canal après paiement (synchrone)"""
        try:
            import asyncio
            return asyncio.run(self._add_member_async(page_id, telegram_user_id, email))
        except Exception as e:
            logger.error(f"Add member sync error: {e}")
            return {'error': str(e)}
    
    async def _add_member_async(self, page_id, telegram_user_id, email=''):
        """Ajoute un membre au canal (async)"""
        try:
            from telegram import Bot
            from services.FirebaseService import firebase_service
            
            # Récupérer la page pour avoir le channel_id
            page = firebase_service.get_landing_page(page_id)
            if not page:
                raise ValueError(f"Page not found: {page_id}")
            
            channel_id = page.get('telegramChannelId') or page.get('telegram', {}).get('channelId')
            if not channel_id:
                raise ValueError("No Telegram channel connected to this page")
            
            bot = Bot(token=self.bot_token)
            
            # Créer lien d'invitation
            expire_date = datetime.now() + timedelta(hours=24)
            invite_link = await bot.create_chat_invite_link(
                chat_id=channel_id,
                expire_date=expire_date,
                member_limit=1
            )
            
            # Envoyer le lien à l'utilisateur
            try:
                await bot.send_message(
                    chat_id=int(telegram_user_id),
                    text=(
                        "🎉 **Paiement confirmé !**\n\n"
                        f"Voici votre lien d'accès au canal (valable 24h):\n"
                        f"{invite_link.invite_link}\n\n"
                        "⚠️ Ce lien est à usage unique."
                    ),
                    parse_mode='Markdown'
                )
                logger.info(f"✅ Invite link sent to user {telegram_user_id}")
            except Exception as send_error:
                logger.error(f"Failed to send message to user: {send_error}")
            
            return {
                'success': True,
                'invite_link': invite_link.invite_link,
                'telegram_user_id': telegram_user_id,
                'channel_id': channel_id
            }
            
        except Exception as e:
            logger.error(f"Add member async error: {e}")
            return {'error': str(e)}
    
    def remove_member_from_channel_sync(self, page_id, telegram_user_id):
        """Retire un membre du canal (synchrone)"""
        try:
            import asyncio
            return asyncio.run(self._remove_member_async(page_id, telegram_user_id))
        except Exception as e:
            logger.error(f"Remove member sync error: {e}")
            return {'error': str(e)}
    
    async def _remove_member_async(self, page_id, telegram_user_id):
        """Retire un membre du canal via Telethon (async)"""
        try:
            from services.FirebaseService import firebase_service
            from telethon import TelegramClient
            
            # Récupérer la page
            page = firebase_service.get_landing_page(page_id)
            if not page:
                raise ValueError(f"Page not found: {page_id}")
            
            channel_id = page.get('telegramChannelId') or page.get('telegram', {}).get('channelId')
            if not channel_id:
                raise ValueError("No Telegram channel connected")
            
            # Utiliser Telethon pour exclure (nécessite userbot)
            if not self.api_id or not self.api_hash:
                raise ValueError("Telethon credentials not configured")
            
            async with TelegramClient('userbot_session', int(self.api_id), self.api_hash) as client:
                await client.start(phone=os.getenv('PHONE_NUMBER'))
                await client.kick_participant(channel_id, int(telegram_user_id))
                
            logger.info(f"✅ User {telegram_user_id} removed from channel {channel_id}")
            
            return {
                'success': True,
                'telegram_user_id': telegram_user_id,
                'channel_id': channel_id
            }
            
        except Exception as e:
            logger.error(f"Remove member async error: {e}")
            return {'error': str(e)}
    
    # ==================== CHANNEL INFO ====================
    
    def get_channel_info_sync(self, channel_id):
        """Récupère les infos d'un canal (synchrone)"""
        try:
            import asyncio
            return asyncio.run(self._get_channel_info_async(channel_id))
        except Exception as e:
            logger.error(f"Get channel info error: {e}")
            return None
    
    async def _get_channel_info_async(self, channel_id):
        """Récupère les infos d'un canal (async)"""
        try:
            from telegram import Bot
            
            bot = Bot(token=self.bot_token)
            chat = await bot.get_chat(chat_id=channel_id)
            
            return {
                'id': chat.id,
                'title': chat.title,
                'username': chat.username,
                'type': chat.type,
                'member_count': getattr(chat, 'member_count', None)
            }
            
        except Exception as e:
            logger.error(f"Get channel info async error: {e}")
            return None


# Instance globale du service
telegram_service = TelegramService()